Add ee_alloc_context (NativeAOT)

noahfalk opened the pull request:
This change is some preparatory refactoring for the randomized allocation sampling feature. We need to add more state onto allocation context but we don't want to do a breaking change of the GC interface. The new state only needs to be visible to the EE but we want it physically near the existing alloc context state for good cache locality. To accomplish this we created a new ee_alloc_context struct which contains an instance of gc_alloc_context within it.

The new field ee_alloc_context::combined_limit should be used by fast allocation helpers to determine when to go down the slow path. Most of the time combined_limit has the same value as alloc_limit, but periodically we need to emit an allocation sampling event on an object that is somewhere in the middle of an AC. Using combined_limit rather than alloc_limit as the slow path trigger allows us to keep all the sampling event logic in the slow path.

There is another PR for CoreCLR making the same change: https://github.com/dotnet/runtime/pull/104849

noahfalk committed:
Add ee_alloc_context (NativeAOT)

This change is some preparatory refactoring for the randomized allocation sampling feature. We need to add more state onto allocation context but we don't want to do a breaking change of the GC interface. The new state only needs to be visible to the EE but we want it physically near the existing alloc context state for good cache locality. To accomplish this we created a new ee_alloc_context struct which contains an instance of gc_alloc_context within it.

The new ee_alloc_context::combined_limit should be used by fast allocation helpers to determine when to go down the slow path. Most of the time combined_limit has the same value as alloc_limit, but periodically we need to emit an allocation sampling event on an object that is somewhere in the middle of an AC. Using combined_limit rather than alloc_limit as the slow path trigger allows us to keep all the sampling event logic in the slow path.

jkotas commented:
> In a future PR we will add a field called combined_limit 

I think you can add it in this PR. Adding the extra field is going to touch a lot of the same lines as this PR.

jkotas commented:
(Also, same comment as https://github.com/dotnet/runtime/pull/104851#issuecomment-2226998081.)

jkotas commented on src/coreclr/nativeaot/Runtime/thread.inl:
```diff
@@ -2,6 +2,36 @@
+    return combined_limit;
+}
+
+// It seems like there is a desire not to include a definition of gc_alloc_context in a more global place within
```
This used to be the case in the early days of Redhawk. We do not care. Feel free to re-arrange things any way you want (bonus points for making things like field names same as in regular coreclr).

noahfalk replied:
Thanks for the info. I'm going to leave this PR as-is to reduce churn but I'll see if this is easy to improve upon in the followup PRs for this feature.

jkotas commented on src/coreclr/nativeaot/Runtime/thread.inl:
```diff
@@ -2,6 +2,36 @@
+    return combined_limit;
+}
+
+// It seems like there is a desire not to include a definition of gc_alloc_context in a more global place within
+// the NativeAOT runtime? Instead some individual files include their own definition as needed and others reference
+// gcinterface.h to get the official definition. This .inl file gets included from multiple places some of which
+// do define the type and others that do not. To avoid getting a redefinition error I added this private definition.
```
```suggestion
// Workaround for https://github.com/dotnet/runtime/issues/96081
```

noahfalk commented on src/coreclr/nativeaot/Runtime/thread.cpp:
```diff
@@ -353,6 +353,7 @@ void Thread::Detach()
     gc_alloc_context* context = GetAllocContext();
     s_DeadThreadsNonAllocBytes += context->alloc_limit - context->alloc_ptr;
     GCHeapUtilities::GetGCHeap()->FixAllocContext(context, NULL, NULL);
+    GetEEAllocContext()->UpdateCombinedLimit();
```
We don't need it for correctness. It was just there for consistency. I'm fine to remove it.

jkotas replied:
I would remove it.

noahfalk replied:
I just pushed a new commit that includes that together with other feedback changes.

noahfalk committed:
PR feedback

- removed unnecessary UpdateCombinedLimit() in thread detach
- updated comment for workaround on 96081
- swapped to updating combined_limit inside GcEnumAllocContexts() instead of in RestartEE()

jkotas commented on src/coreclr/nativeaot/Runtime/gcenv.ee.cpp:
```diff
@@ -146,7 +136,17 @@ void GCToEEInterface::GcEnumAllocContexts(enum_alloc_context_func* fn, void* par
+        // combined_limit up-to-date. Note that the GC has multiple threads running this enumeration concurrently
+        // with no synchronization. If you need to change this code think carefully about how that concurrency
+        // may affect the results.
+        if(ac->alloc_limit == 0)
```
```suggestion
        if (ac->alloc_limit == 0)
```

noahfalk committed:
Update src/coreclr/nativeaot/Runtime/gcenv.ee.cpp

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

noahfalk committed:
Reduce unnecessary combined_limit writes

noahfalk commented:
@jkotas @MichalStrehovsky - I've applied the feedback from the CoreCLR issue to this one as well. Is anything else needed before you can signoff?

jkotas commented:
/azp run runtime-nativeaot-outerloop

jkotas commented on src/coreclr/nativeaot/Runtime/DebugHeader.cpp:
```diff
@@ -163,6 +163,9 @@ extern "C" void PopulateDebugHeaders()
     MAKE_DEBUG_FIELD_ENTRY(dac_gc_heap, finalize_queue);
     MAKE_DEBUG_FIELD_ENTRY(dac_gc_heap, generation_table);
 
+    MAKE_SIZE_ENTRY(ee_alloc_context);
```
@mikem8361 Is this going to need matching changes in SOS for native AOT? Do we need to bump the data contract version to make it work?

mikem8361 replied:
Yes, this will require changes to the DCR/NativeAOT SOS and a minor version bump is needed.

noahfalk replied:
I think we'd want a major version bump because of the change to RuntimeThreadLocals below right? Code that expect gc_alloc_context to exist at RuntimeThreadLocals.m_rgbAllocContextBuffer isn't going to find that field any longer and anything that hardcoded the offset no longer has the right offset.

noahfalk replied:
I went ahead and bumped the major version

jkotas commented:
`nativeaot-outerloop (Build windows-x64 Release NativeAOT_Checked` timeouts look like infrastructure issues, but they do not go away with reruns. Either the Helix queues are overloaded, or there is a bug that makes the test run a lot slower.

jkotas reviewed (commented):
Looks good to me otherwise

noahfalk commented on src/coreclr/nativeaot/Runtime/thread.h:
```diff
@@ -83,9 +83,43 @@ struct InlinedThreadStaticRoot
+    //   of sampled bytes didn't select a byte in this AC.
+    // - if combined_limit < alloc_limit there is a sample limit in the AC. sample_limit = combined_limit.
+    uint8_t* combined_limit;
+    uint8_t m_rgbAllocContextBuffer[SIZEOF_ALLOC_CONTEXT];
```
To confirm, you would prefer this in the NativeAOT?
```
_combined_limit
_rgbAllocContextBuffer
```

I'm happy to do any style - its mostly a question of where we want the discontinuties to be. The current approach started with snake case gc_alloc_context fields in CoreCLR, then snake case ee_alloc_context fields, then duplicating that within NativeAOT. With this approach we'll have a CoreCLR<->NativeAOT difference in the names (which I have no objection to).

jkotas replied:
Ideally, the names would be the same in both CoreCLR and NativeAOT, and use consistent naming convention.

(I have not looked into it in details. I wrote this comment based on the local observation where two fields next to each other use very different naming style.)

noahfalk replied:
Unless we change the style of pre-existing code there is no option that is simultaneously consistent within types, across types used together, and across runtimes. At best if we change all the new fields to use m_ with PascalCasing we can be consistent on 2 out of 3. The names would not be consistent with field naming used in GC so we'll have expressions like `GetEEAllocContext()->m_gcAllocationContext.alloc_ptr`. If that sounds good I will make the change.

jkotas replied:
I do not have a strong opinion. The more consistent and unified, the better. Also, I believe `_` prefix is the most prevalent style in the repo, so I would lean towards that when having a choice.

noahfalk commented:
I've updated for the feedback so far. Everything is addressed except NativeAOT SOS which has to be handled out-of-band. Looking at the time I think I may need to call it quits on this for .NET 9 and instead treat it as .NET 10 on a slower cadence. At this point I don't think I have more time available right now to do a staged checkin, revalidate testing after the ongoing changes, and continue to have time in reserve to respond to potential post-checkin issues.

noahfalk committed:
Bumping the debug header major version
